ClusterClass and managed topologies blog (#29763)
kubernetes/website: content/en/blog/_posts/2021-09-21-clusterclass-and-managed-topologies.md

---
layout: blog

> title: "Introducing ClusterClass and managed topologies in Cluster API"

Reviewer: Could you please convert this to title case for consistency, that is, `Introducing ClusterClass and Managed Topologies in Cluster API` 😄

           ref:
             apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
             kind: KubeadmConfigTemplate
             name: linux-bootstrap
         infrastructure:
           ref:
             apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
             kind: VSphereMachineTemplate
             name: linux-vsphere-template
     - class: windows-worker
       template:
         bootstrap:
           ref:
             apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
             kind: KubeadmConfigTemplate
             name: windows-bootstrap
         infrastructure:
           ref:
             apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
             kind: VSphereMachineTemplate
             name: windows-vsphere-template
   infrastructure:
     ref:
       apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
       kind: VSphereClusterTemplate
       name: vsphere-cluster
```

The possibilities are endless; you can get a default Cluster Class from the community, “off-the shelf” classes from your vendor of choice, “certified” classes from the platform admin in your company, or even create custom ones for advanced scenarios. 

## Managed Topologies

> Managed topologies is how you put the power that ClusterClass provides into action.

Reviewer 2: I think I would choose:
```
Managed topologies let you put the power that ClusterClass provides into action.
```

> Managed topologies is how you put the power that ClusterClass provides into action.

Reviewer 3: :+1: that wording looks good too.

---
layout: blog
title: "Introducing ClusterClass and Managed Topologies in Cluster API"
date: 2021-09-21
slug: capi-clusterclass-and-managed-topologies
---

**Author:** Fabrizio Pandini (VMware)

The [Cluster API community](https://cluster-api.sigs.k8s.io/) is happy to announce the implementation of *ClusterClass and Managed Topologies*, a new feature that will greatly simplify how you can provision, upgrade, and operate multiple Kubernetes clusters in a declarative way.

## A little bit of context…

Before getting into the details, let's take a step back and look at the history of Cluster API.

> The [Cluster API project](https://github.com/kubernetes-sigs/cluster-api/) started three years ago, and the first releases focused on extensibility and implementing a declarative API that allows a seamless experience across infrastructure providers. This was a success with many cloud providers: AWS, Azure, vSphere, Metal3, Digital Ocean, GCP and still counting.

Reviewer 3: nit: CNCF policy for listing 3rd party products and / or vendors is to use either alphabetical order, or a random shuffle.
How about:
```suggestion
The [Cluster API project](https://github.com/kubernetes-sigs/cluster-api/) started three years ago, and the first releases focused on extensibility and implementing a declarative API that allows a seamless experience across infrastructure providers. This was a success with many cloud providers: AWS, Azure, Digital Ocean, GCP, Metal3, vSphere…  and still counting.
```

Reviewer 2: @sftim Out of curiosity: how do you now something uses "random shuffle"? :)

> With extensibility addressed, the focus shifted to features, like automatic control plane and etcd management, health based machine remediation, machine rollout strategies and more.

Reviewer 2: ```suggestion
With extensibility addressed, the focus shifted to features, like automatic control plane and etcd management, health-based machine remediation, machine rollout strategies and more.
```
nit (not 100% sure)

Fast forwarding to 2021, with lots of companies using Cluster API to manage fleets of Kubernetes clusters running workloads in production, the community focused its effort on stabilization of both code, APIs and documentation, and on extensive test signals which inform Kubernetes releases.

With solid foundations in place, and a vibrant and welcoming community that still continues to grow, it was time to plan another iteration on our UX for both new and advanced users.

Enter ClusterClass and manage topologies, tada!

## ClusterClass

As the name suggests, ClusterClass and managed topologies are built in two parts.

The idea behind ClusterClass is simple: let’s define the shape of your cluster once, and then reuse it many times, abstracting the complexities and the internals of a Kubernetes cluster away.

![Defining a ClusterClass](/images/blog/2021-09-21-clusterclass-and-managed-topologies/clusterclass.svg)

ClusterClass, at its heart, is a collection of Cluster and Machine templates. You can use it as a “stamp” that can be leveraged to create many clusters of a similar shape.

```yaml
---

> apiVersion: cluster.x-k8s.io/v1alpha4

Reviewer 2: @fabriziopandini Can we / Do we want to use our new v1beta1 types here? :)
(I guess depends on timing of v1beta1 / this blog post)

       ref:
         apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
         kind: VSphereMachineTemplate
         name: controlplane-vsphere-machinetemplate
   workers:
     deployments:
     - class: linux-worker
       template:
         bootstrap:
           ref:
             apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
             kind: KubeadmConfigTemplate
             name: linux-bootstrap
         infrastructure:
           ref:
             apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
             kind: VSphereMachineTemplate
             name: linux-vsphere-template
     - class: windows-worker
       template:
         bootstrap:
           ref:
             apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
             kind: KubeadmConfigTemplate
             name: windows-bootstrap
         infrastructure:
           ref:
             apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
             kind: VSphereMachineTemplate
             name: windows-vsphere-template
   infrastructure:
     ref:
       apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
       kind: VSphereClusterTemplate
       name: vsphere-cluster
```

> The possibilities are endless; you can get a default Cluster Class from the community, “off-the shelf” classes from your vendor of choice, “certified” classes from the platform admin in your company, or even create custom ones for advanced scenarios.

Reviewer 2: ```suggestion
The possibilities are endless; you can get a default ClusterClass from the community, “off-the shelf” classes from your vendor of choice, “certified” classes from the platform admin in your company, or even create custom ones for advanced scenarios.
```
nit
